### DISCUSSION TITLE
Node draining (for in-place Kubernetes upgrade)

### DISCUSSION THREAD

**Discussion by ehemmerlin** [longhorn/longhorn#8075] [Q&A] [answered]
When draining a node after having cordoned it as stated in the [documentation](https://longhorn.io/docs/1.6.0/maintenance/maintenance/#updating-the-node-os-or-container-runtime) we get the following error:
`error when evicting pods/"instance-manager-xxx" -n "longhorn-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.`

Having the allowed disruptions of the instance manager pods set to 0 seems to cause the issue:
![image](https://github.com/longhorn/longhorn/assets/4458966/3792f38c-e473-4220-bc15-95d9e285af64)

How is it possible to drain a node containing Longhorn on it? Is there a value related to the pod disruption budget we can update in the Helm chart? We need this for in-place Kubernetes upgrades.

This anwser is close to what we need, but involves manual steps that are not possible in an Ansible automated way of upgrading Kubernetes: https://github.com/longhorn/longhorn/discussions/4102#discussioncomment-3118777

We use Longhorn version 1.5.3 installed using helm chart.

**Comment by derekbit** [comment]
>  involves manual steps

Do you mean `Evict all the replicas from the node via Longhorn UI` is manual step?
If so, you can directly edit node.longhorn.io resource instead of updating it via Longhorn UI.

**Reply by ehemmerlin** [reply]
Thank you for your answer. In the meantime we discovered that all replicas have to be healthy in order for instance managers to allow the node to be drained. We now need a way to check each pvc state or replicas state before triggering a drain on nodes, which leads to this post: https://github.com/longhorn/longhorn/discussions/8103

**Reply by ehemmerlin** [reply, marked as answer]
For the record, checking the health of Longhorn volumes before draining a node ensures that instance manager will allow the node to be drained. This Bash script checks the state of Longhorn volumes by querying Longhorn CRD using kubectl.

```
nbr_of_non_healthy_volumes=$(kubectl get volume -n longhorn-system -o=jsonpath="{range .items[*]}{.status.robustness} {.status.state}{\"\n\"}" | grep attached | awk {'print $1'} | grep -v healthy | wc -l)
while [ $nbr_of_non_healthy_volumes -gt 0 ]; do sleep 3; nbr_of_non_healthy_volumes=$(kubectl get volume -n longhorn-system -o=jsonpath="{range .items[*]}{.status.robustness} {.status.state}{\"\n\"}" | grep attached | awk {'print $1'} | grep -v healthy | wc -l); done
```
